### ISSUE TITLE
ADBD crashes often

### ISSUE THREAD

**Issue by heckarman** [GrapheneOS/os-issue-tracker#2724]
I've Shizuku installed and often times adbd crashes forcing Shizuku to shut it's running services: 
```
type: crash
osVersion: google/bluejay/bluejay:14/UP1A.231105.003/2023112600:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 6847s

abortMessage: failed to delete fd 24 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+164, pc 62e64)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+12, pc 368bc)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80, pc 38730)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+352, pc 37c30)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+1620, pc c5154)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, init_jdwp()::$_0> >(void*)+168, pc c3928)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+204, pc cfa0c)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64, pc 64730)
```

**Comment by thestinger** [comment]
Is this with the unmodified official releases of GrapheneOS?

**Comment by heckarman** [comment]
Yes. I've had bootloader unlocked for experiments apart from that, nothing's changed.

**Comment by thestinger** [comment]
It looks like the OS is modified.

**Comment by thestinger** [comment]
Please try with a fresh install. This looks like adb is trying to use root access which isn't in production builds.

It's meant to be locked in regular use which gives strong assurance that it's not modified via verified boot.

**Comment by heckarman** [comment]
Apart from boot image nothing was modified until yesterday. Now that, apart from the fact that bootloader is unlocked, nothing is modified. Will look into it.

I hope that GrapheneOS will prioritize data backup solution ASAP so that I can trust the OS with my data.

**Comment by muhomorr** [comment]
This is not a new crash, it just wasn't visible in the UI before. Next release will disable these notifications by default.

**Comment by heckarman** [comment]
> This is not a new crash, it just wasn't visible in the UI before. Next release will disable these notifications by default.

Oh wait! I had enabled `Always show crash dialogue` in developer settings.

Getting context now. But may I know why it occurs? I'm just curious now.

**Comment by David-Prock** [comment]
My boot loader is locked and i dont have that package that the other person in post mentioned.
And i got this error.
In fact i have been getting lots of different errors

**Comment by elvisisvan** [comment]
having a similar error: 
```
type: crash
osVersion: google/bluejay/bluejay:14/UP1A.231105.003/2023112900:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 6852s

abortMessage: failed to delete fd 20 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+164, pc 62e64)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+12, pc 368bc)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80, pc 38730)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+352, pc 37c30)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+1620, pc c5154)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, init_jdwp()::$_0> >(void*)+168, pc c3928)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+204, pc cf9ec)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64, pc 64730)
```

**Comment by Geo25rey** [comment]
TL;DR

`adbd` crashes on multiple devices some number of hours after being used on both Pixel 6a (bluejay) and Pixel 7a (lynx) devices and both rooted and not rooted systems. It's very likely to be the same or similar issue since the error happens at `adbconnection_listen(void (*)(int, ProcessInfo))+1620, pc c5154` on all devices.

Also, my stacktrace from Pixel 7a (lynx):
```
type: crash
osVersion: google/lynx/lynx:14/UP1A.231105.003/2023112900:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 125806s

abortMessage: failed to delete fd 20 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+164, pc 62e64)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+12, pc 368bc)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80, pc 38730)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+352, pc 37c30)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+1620, pc c5154)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, init_jdwp()::$_0> >(void*)+168, pc c3928)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+204, pc cf9ec)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64, pc 64730)
```

**Comment by Geo25rey** [comment]
@thestinger Do you know where to find this adbconnection_listen() function? The [last commit](https://github.com/GrapheneOS/platform_art/commit/bd396c001fd3b13a7451876ec8c535bb597ea350) that I can find referencing it is from 2020 and it deletes the file it was in without replacing it. I assume it went to another repo, but it's hard to find out of hundreds of repos.

**Comment by chenxiaolong** [comment]
It was moved to https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/adb/libs/adbconnection/adbconnection_server.cpp;l=37;drc=fb529d52f4485c1fdbe6208f20af6578493e1ca7

Looks like GrapheneOS just uses the upstream AOSP repo for that.

**Comment by thestinger** [comment]
The current stable release tag for AOSP is android-14.0.0_r14 which is the basis for GrapheneOS and the stock Pixel OS, not the `main` branch.

**Comment by thestinger** [comment]
The latest tag was android-14.0.0_r14 not android-14.0.0_r11. The latest tag is now android-14.0.0_r16 with today's release.

**Comment by thestinger** [comment]
> I've seen your name on a number of commit regarding ADB over the past few years. What are your insights on what and why this might be happening?

I don't know what you mean.

**Comment by Geo25rey** [comment]
> > I've seen your name on a number of commit regarding ADB over the past few years. What are your insights on what and why this might be happening?
> 
> I don't know what you mean.

9 commits you made [here](https://github.com/search?q=org%3AGrapheneOS++author%3Athestinger+adb&type=commits&s=author-date&o=desc) and another 3 [here](https://github.com/search?q=org%3AGrapheneOS++author%3Athestinger+adbd&type=commits&s=author-date&o=desc) earliest from 2016.

You seem much more knowledgeable than me on ADB's innerworkings. I figured you would have some more insight than I do after working on ADB related things for 7 years.

So, what would you suggest to debug this issue?

**Comment by thestinger** [comment]
Those aren't changes to the ADB code.

**Comment by whatthesamuel** [comment]
Same here: latest graphene unmodified/pixel 6a

**Comment by e-t-l** [comment]
Pixel 7 running unmodified GOS stable channel.

Maybe worth noting that the "fd [number]" in the line `failed to delete fd 31` changes each time I get this crash, and also appears to have no consistency with the crash reports shared by others.

So it seems like something might be wrong with the file descriptors? (I have no knowledge of android kernel or adbd code, so i may be totally wrong here)

```
type: crash
osVersion: google/panther/panther:14/AP1A.240505.005/2024052100:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 157841s

abortMessage: failed to delete fd 31 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+164, pc 64e84)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+12, pc 17c6c)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80, pc 19ad0)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+352, pc 18fd0)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+1584, pc 118e40)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, init_jdwp()::$_0> >(void*)+168, pc 1162d8)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+204, pc cf93c)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64, pc 66730)
```

**Comment by MatejKovacic** [comment]
I have similar issue, have locked bootloader and unmodified GrapheneOS on Pixel 6a. I have enabled developer options, but did not change any settings.

**Comment by andmagdo** [comment]
Exact same issue, pixel 6 running newest stable channel update, unmodified grapheneOS, locked bootloader, shizuku not running, USB debugging on but since the device booted up, it hasn't been connected to any computers
```
type: crash
osVersion: google/oriole/oriole:14/AP2A.240705.004/2024070201:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 65239s

abortMessage: failed to delete fd 18 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+164, pc 64a24)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+12, pc 15e3c)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80, pc 17ca0)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+352, pc 171a0)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+1568, pc 1166b0)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, init_jdwp()::$_0> >(void*)+168, pc 113b58)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+204, pc 7679c)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64, pc 67d50)
```

**Comment by TimoGoetze** [comment]
after upgrade to android 15  on oct 20, adb crash for me, too.
```
type: crash
osVersion: google/cheetah/cheetah:15/AP3A.241005.015/2024102100:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 45970s

abortMessage: failed to delete fd 18 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+164, pc 63c84)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+12, pc 12abc)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80, pc 14990)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+512, pc 13ef0)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+2340, pc d2674)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy[abi:nn180000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, init_jdwp()::$_0>>(void*)+168, pc ccec8)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+200, pc 76198)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64, pc 66eb0)`
```

**Comment by pixincreate** [comment]
this has become too frequent with A15

**Comment by pixincreate** [comment]
I use Shizuku in my daily workflow for multiple stuff. And adbd crashes too frequently (like 15 - 20 times a day)

**Comment by pixincreate** [comment]
App behavior might have been changed on Android 15 is what you meant? In that case I guess I'll spend time trying to figure out.

**Comment by tmdobbin** [comment]
Just would like to say this is happening to me really often. I'll include the crash log here too if its helpful, but it doesnt seem much different from the others here. 

```
type: crash
osVersion: google/akita/akita:15/AP3A.241005.015/2024102400:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 18917s

abortMessage: failed to delete fd 19 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control
MTE: enabled

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+168, pc 67d78)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+16, pc 14c10)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+84, pc 16b94)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+516, pc 160b4)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+2340, pc d6394)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy[abi:nn180000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, init_jdwp()::$_0>>(void*)+172, pc d09ec)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+200, pc 7afe8)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+68, pc 6b164)
```

**Comment by pixincreate** [comment]
> Yes, the apps might be behaving differently and doing something which is more likely to trigger this existing bug which isn't new.

The issue seems to be something else...

I do not have Shizuku enabled as it crashed and I'm too far away from WiFi.
It still crashed 🫤:

```
type: crash
flags: dev options enabled
osVersion: google/bluejay/bluejay:15/AP3A.241105.007/2024111800:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 4359s

abortMessage: failed to delete fd 19 from JDWP epoll fd: Bad file descriptor

signal: 6 (SIGABRT), code -1 (SI_QUEUE)
threadName: jdwp control

backtrace:
    /apex/com.android.runtime/lib64/bionic/libc.so (abort+164, pc 63cc4)
    /apex/com.android.adbd/lib64/libbase.so (android::base::DefaultAborter(char const*)+12, pc 12abc)
    /apex/com.android.adbd/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80, pc 14990)
    /apex/com.android.adbd/lib64/libbase.so (android::base::LogMessage::~LogMessage()+512, pc 13ef0)
    /apex/com.android.adbd/bin/adbd (adbconnection_listen(void (*)(int, ProcessInfo))+2340, pc d2674)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy[abi:nn180000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, init_jdwp()::$_0>>(void*)+168, pc ccec8)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+200, pc 761c8)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64, pc 66ef0)
```

**Comment by shenki** [comment]
Crash when attempting to connect to device using adb over TCP, without pairing first. The crash appears to happen on the second connection attempt.

Pairing allows a successful connection without crashing.

```
type: crash
flags: dev options enabled
osVersion: google/caiman/caiman:15/AP4A.250105.002/2025011500:user/release-keys
uid: 2000 (u:r:adbd:s0)
cmdline: /apex/com.android.adbd/bin/adbd --root_seclabel=u:r:su:s0
processUptime: 106s

signal: 11 (SIGSEGV), code 9 (SEGV_MTESERR), faultAddr d00c46bd1e08c28
threadName: adbd
MTE: enabled

backtrace:
    /apex/com.android.adbd/bin/adbd (Connection::Serial() const+16, pc c2470)
    /apex/com.android.adbd/bin/adbd (BlockingConnectionAdapter::Stop()+820, pc c2b44)
    /apex/com.android.adbd/bin/adbd (atransport::Kick()+124, pc c44ac)
    /apex/com.android.adbd/bin/adbd (void* std::__1::__thread_proxy[abi:nn190000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, adbd_auth_tls_handshake(atransport*)::$_0>>(void*)+360, pc cddc8)
    /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+132, pc 7ba94)
    /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+68, pc 6b474)
```

**Comment by chenxiaolong** [comment]
I found the root cause for this bug (`-EBADF` when unregistering jwdp connection fd from epoll) and submitted a PR with the fix at: https://github.com/GrapheneOS/platform_packages_modules_adb/pull/3

**Comment by Alexander-Shukaev** [comment]
Still happening even after today's update...

**Comment by chenxiaolong** [comment]
@Alexander-Shukaev Please provide the new logs from the crash. It's unlikely that it's the same `EBADF` crash from before.

**Comment by chenxiaolong** [comment]
@thestinger Thanks for the link! Yep, I can definitely help review https://github.com/GrapheneOS/platform_packages_modules_adb/pull/4 after work today.

Sure, I'll DM you as soon as I can.

**Comment by chenxiaolong** [comment]
![Image](https://github.com/user-attachments/assets/4b10d881-188a-44eb-91a1-def64b6d1113)
